data.lua

```diff
@@ -1,81 +1,202 @@
-
 require 'os'
 require 'paths'
 require 'torch'
 require 'sys'
-require 'shared'
 local lfs = require "lfs"
 
--- shamelessly adapted from Karpathy's char-rnn :-)
-function text_to_t7(in_textfile, out_tensorfile, out_vocabfile)
-  local cache_len = 10000
-  local rawdata
-  local tot_len = lfs.attributes (in_textfile, "size")
-  print('tot_len', tot_len)
+-- Modified from https://github.com/oxford-cs-ml-2015/practical6
+-- the modification included support for train/val/test splits
+-- Modified from https://github.com/karpathy/char-rnn  (MIT License)
+-- modifications include name change and use of torch.class (not many)
 
-  local input_coded = torch.ByteTensor(tot_len)
-  local vocab = {}
-  local ivocab = {}
-  local state = {}
-  state.vocab = vocab
-  state.ivocab = ivocab
-
---  -- add a null character, which will be used for start-of-sequence
---  ivocab[1] = 0
---  vocab[0] = 1
-
-  local f = io.open(in_textfile, "r")
-  local input_string = f:read(cache_len)
-  local pos = 1
-  repeat
-    local len = input_string:len()
-    for i=1,len do
-      local char = input_string:byte(i)
-      if vocab[char] == nil then
-        ivocab[#ivocab + 1] = char
-        vocab[char] = #ivocab
-      end
-      local v = vocab[char]
-      input_coded[pos] = v
-      pos = pos + 1
+local CharTextLoader = torch.class("nn.CharTextLoader")
+
+CharTextLoader.TRAIN = 1
+CharTextLoader.VAL = 2
+CharTextLoader.TEST = 3
+
+function CharTextLoader:__init(data_dir, batch_size, seq_length, split_fractions)
+  -- split_fractions is e.g. {0.9, 0.05, 0.05} (train, valid, test)
+  
+  -- train / val / test split for data, in fractions
+  split_fractions[3] = split_fractions[3] or math.max(0, 1 - (split_fractions[1] + split_fractions[2]))
+
+  local input_file = path.join(data_dir, 'input.txt')
```

Comment:
`path` is implemented in `th`, not available in `luajit`, as far as I know?  I like to be able to run things in luajit (in fact, almost always do), because it's much easier to run luajit from gdb.  At least, I know of no way of getting `th` to run ok in gdb?  (Also, luajit tends to give much more useful error messages, if c/c++ code crashes)